src/rt/sched/multimessage.h

```diff
@@ -14,29 +14,61 @@ namespace verona::rt
 
   class MultiMessage
   {
-    struct MultiMessageBody
+    /**
+     * This represents a message that is send to a behaviour.
+     *
+     * The layout requires that after the allocation there is space for
+     * `count` cown pointers, and then the behaviour's body.
+     *
+     * This layout allows the message body to be single allocation even though
+     * there are multiple different sized pieces.
+     */
+    struct Body
     {
-      size_t index;
-      size_t count;
-      Cown** cowns;
       std::atomic<size_t> exec_count_down;
-      Behaviour* behaviour;
+      size_t count;
+
+    private:
+      Body(size_t count) : exec_count_down(count), count(count) {}
+
+    public:
+      Cown** get_cowns_array()
+      {
+        return snmalloc::pointer_offset<Cown*>(this, sizeof(Body));
+      }
+
+      Behaviour* get_behaviour()
+      {
+        return snmalloc::pointer_offset<Behaviour>(
+          this, sizeof(Body) + sizeof(Cown*) * count);
+      }
+
+      /**
+       * Allocates a message body with sufficient space for the
+       * cowns_array and the behaviour.  Neither the cowns array
+       * or the behaviour are initialised by this function.
+       */
+      static Body* make(Alloc& alloc, size_t count, size_t behaviour_size)
+      {
+        size_t size = sizeof(Body) + (sizeof(Cown*) * count) + behaviour_size;
+        return new (alloc.alloc(size)) Body(count);
+      }
     };
 
   private:
-    MultiMessageBody* body;
+    // The body of the actual message.
+    Body* body;
     friend verona::rt::MPSCQ<MultiMessage>;
     friend class Cown;
 
     std::atomic<MultiMessage*> next{nullptr};
 
-    inline MultiMessageBody* get_body()
+    inline Body* get_body()
```

Comment:
This can return `nullptr`, but it probably should be an assertion failure if it would.  Can we make it return a `Body&` and add `assert((Body*)((uintptr_t)body & ~Object::MARK_MASK) != nullptr)`?

Reply:
So I looked at this change, but we currently do
```
alloc.dealloc(m->get_body());
```
and I didn't really like that with the reference. It felt wrong.

Reply:
I think it feels wrong because it is wrong: it leaves body as a dangling pointer.  We should be deallocating and nulling out the pointer at the same time, possibly with a `clear_body` method.

Reply:
I agree with David on a more dedicated method. I am just considering if there is an argument to consider how ownership flows outside of this class. For example, should ownership flow to `Scheduler::local()->message_body`?

Reply:
I think this is a more complex refactor than I want in this PR. I will create a second PR address this nicely.